dev/sparktestsupport/modules.py

```diff
@@ -58,14 +95,19 @@ def __init__(self, name, dependencies, source_file_regexes, build_profile_flags=
             is not explicitly changed.
         :param should_run_r_tests: If true, changes in this module will trigger all R tests.
         :param should_run_build_tests: If true, changes in this module will trigger build tests.
+        :param python_test_paths: A set of python test paths to be discovered
+        :param python_excluded_tests: A set of excluded Python tests
         """
         self.name = name
         self.dependencies = dependencies
         self.source_file_prefixes = source_file_regexes
         self.sbt_test_goals = sbt_test_goals
         self.build_profile_flags = build_profile_flags
         self.environ = environ or {}
-        self.python_test_goals = python_test_goals
+        discovered_goals = _discover_python_unittests(python_test_paths)
+        # Final goals = Manual specified goals + Discoverd goals - Excluded goals
+        all_goals = set(python_test_goals) | set(discovered_goals)
+        self.python_test_goals = sorted(list(set(all_goals) - set(python_excluded_tests)))
```

Comment:
@Yikun what about we add a private variable at these test slow modules, and detect them by checking if that variable is defined?

For example,

```diff
diff --git a/python/pyspark/pandas/tests/test_dataframe.py b/python/pyspark/pandas/tests/test_dataframe.py
index e54b7835c22..22f2b5da93e 100644
--- a/python/pyspark/pandas/tests/test_dataframe.py
+++ b/python/pyspark/pandas/tests/test_dataframe.py
@@ -50,6 +50,10 @@ from pyspark.testing.sqlutils import SQLTestUtils
 from pyspark.pandas.utils import name_like_string


+# This is used in run-tests.py blah blah. see modules.py blah blah
+_slow_test = True
+
+
 class DataFrameTest(PandasOnSparkTestCase, SQLTestUtils):
     @property
     def pdf(self):
```

and then we can filter like:

```python
is_slow_test = hasattr(module, "_slow_test")
```

In this way, we can remove `pyspark_pandas_slow_unittests` define here too.

Also, can we remove `python_test_paths` and `python_excluded_tests` here too? 

We could run `_discover_python_unittests` and directly assign to  `python_test_goals` I guess. For example,

```python
python_test_goals = [
    ...
] + _discover_python_unittests("pyspark/tests")
```

Reply:
Yes, great suggestion, it makes module implementation more clean.

If we want to **only** discover the slow tests
```
python_test_goals = [
    ...
] + _discover_python_unittests("pyspark/tests", discover_slow=True)
```
otherwise will skip slow tests.

I complete it in my next commit. : )